overlord/managers_test.go

```diff
@@ -10021,6 +10022,440 @@ func (s *mgrsSuite) TestUpdateKernelBaseSingleRebootKernelUndo(c *C) {
+		CurrentKernels:         []string{"pc-kernel_1.snap"},
+		CurrentRecoverySystems: []string{"1234"},
+		GoodRecoverySystems:    []string{"1234"},
+
+		Model:          model.Model(),
+		BrandID:        model.BrandID(),
+		Grade:          string(model.Grade()),
+		ModelSignKeyID: model.SignKeyID(),
+	}
+	err = m.WriteTo("")
+	c.Assert(err, IsNil)
+	c.Assert(s.o.DeviceManager().ReloadModeenv(), IsNil)
+
+	pcKernelYaml := "name: pc-kernel\nversion: 1.0\ntype: kernel"
+	baseYaml := "name: core20\nversion: 1.0\ntype: base"
+	siKernel := &snap.SideInfo{RealName: "pc-kernel", SnapID: fakeSnapID("pc-kernel"), Revision: snap.R(1)}
+	snaptest.MockSnap(c, pcKernelYaml, siKernel)
+	siBase := &snap.SideInfo{RealName: "core20", SnapID: fakeSnapID("core20"), Revision: snap.R(1)}
+	snaptest.MockSnap(c, baseYaml, siBase)
+	siSnapd := &snap.SideInfo{RealName: "snapd", Revision: snap.R(1), SnapID: fakeSnapID("snapd")}
+	snaptest.MockSnap(c, "name: snapd\ntype: snapd\nversion: 123", siSnapd)
+	siGadget := &snap.SideInfo{RealName: "pc", Revision: snap.R(1), SnapID: fakeSnapID("pc")}
+	snaptest.MockSnapWithFiles(c, snapYamlGadget, siGadget, [][]string{
+		{"meta/gadget.yaml", pcGadgetYaml},
+	})
+
+	// test setup adds core, get rid of it
+	snapstate.Set(st, "core", nil)
+	snapstate.Set(st, "pc-kernel", &snapstate.SnapState{
+		Active:   true,
+		Sequence: []*snap.SideInfo{siKernel},
+		Current:  snap.R(1),
+		SnapType: "kernel",
+	})
+	snapstate.Set(st, "core20", &snapstate.SnapState{
+		Active:   true,
+		Sequence: []*snap.SideInfo{siBase},
+		Current:  snap.R(1),
+		SnapType: "base",
+	})
+	snapstate.Set(st, "snapd", &snapstate.SnapState{
+		Active:   true,
+		Sequence: []*snap.SideInfo{siSnapd},
+		Current:  snap.R(1),
+		SnapType: "snapd",
+	})
+	snapstate.Set(st, "pc", &snapstate.SnapState{
+		Active:   true,
+		Sequence: []*snap.SideInfo{siGadget},
+		Current:  snap.R(1),
+		SnapType: "gadget",
+	})
+
+	p, _ := s.makeStoreTestSnap(c, pcKernelYaml, "2")
+	s.serveSnap(p, "2")
+	p, _ = s.makeStoreTestSnap(c, baseYaml, "2")
+	s.serveSnap(p, "2")
+	p, _ = s.makeStoreTestSnap(c, "name: snapd\ntype: snapd\nversion: 123", "2")
+	s.serveSnap(p, "2")
+	p, _ = s.makeStoreTestSnapWithFiles(c, snapYamlGadget, "2", [][]string{
+		{"meta/gadget.yaml", pcGadgetYaml},
+	})
+	s.serveSnap(p, "2")
+
+	affected, tss, err := snapstate.UpdateMany(context.Background(), st, []string{"pc-kernel", "core20", "pc", "snapd"}, 0, nil)
+	c.Assert(err, IsNil)
+	c.Assert(affected, DeepEquals, []string{"core20", "pc", "pc-kernel", "snapd"})
+	chg := st.NewChange("update-many", "...")
+	for _, ts := range tss {
+		// skip the taskset of UpdateMany that does the
+		// check-rerefresh, see tsWithoutReRefresh for details
+		if ts.Tasks()[0].Kind() == "check-rerefresh" {
+			c.Logf("skipping rerefresh")
+			continue
+		}
+		chg.AddAll(ts)
+	}
+	return bloader, chg
+}
+
+func (s *mgrsSuite) TestUpdateKernelBaseSingleRebootWithGadgetWithExplicitBase(c *C) {
+	const pcGadget = `
```

Comment:
this probably should have a comment that in this scenario we revert back to having two reboots, that's make it easier to follow the code below that deals with simulating the corresponding restarts

Reply:
added